pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -808,21 +808,31 @@ func (m *ManagerImpl) filterByAffinity(podUID, contName, resource string, availa
 	var withoutTopology []string
 	for d := range available {
 		// Since the same device may be associated with multiple NUMA Nodes. We
-		// need to be careful not to add each device to multiple lists. The
-		// logic below ensures this by breaking after the first NUMA node that
-		// has the device is encountered.
+		// need to be careful not to add each device to multiple lists. The logic
+		// below ensures this by loop throug all NUMA Nodes and select a target
+		// list from the three devices list defined above and then insert the device
+		// into this list. For NUMA node that has the device, this loop will break
+		// only when it found that this node meets the affinity requirement and
+		// set fromAffinity as target list. Otherwise it will keep looping and:
+		//    i: set nodeWithoutTopology as target list when targetList is nil and
+		//   ii: set notFromAffinity as target list when target is nil or nodeWithoutTopology
+		var targetList *[]string
 		for _, n := range nodes {
 			if perNodeDevices[n].Has(d) {
-				if n == nodeWithoutTopology {
-					withoutTopology = append(withoutTopology, d)
+				if n == nodeWithoutTopology && targetList == nil {
+					targetList = &withoutTopology
 				} else if hint.NUMANodeAffinity.IsSet(n) {
-					fromAffinity = append(fromAffinity, d)
-				} else {
-					notFromAffinity = append(notFromAffinity, d)
+					targetList = &fromAffinity
+					break
+				} else if targetList != &notFromAffinity {
+					targetList = &notFromAffinity
 				}
-				break
 			}
 		}
+
+		if targetList != nil {
+			*targetList = append(*targetList, d)
```

Comment:
Do we have unit tests which exercise this flow and highlight the bug? If not, let's add them